furusiya/model/maps/generators/dungeon_generator.py

```diff
@@ -0,0 +1,221 @@
+import math
+from random import randint
+import random
+
+from attrdict import AttrDict
+
+import colors
+from constants import GROUND_CHARACTER, WALL_CHARACTER, color_light_ground, color_dark_ground, color_dark_wall, color_light_wall
+from main_interface import Game
+from model.config import config
+from model.components.walkers.random_walker import RandomWalker
+from model.item_callbacks import cast_heal, cast_lightning, cast_fireball, cast_confuse
+from model.rect import Rect
+from model.factories import monster_factory, item_factory
+
+
+class DungeonGenerator:
+    """
+    Generates a multi-room dungeon map in entirety, by mutating map tiles.
+    Includes population of monsters, map items, etc.
+    """
+    NUM_ITEMS = (10, 20)
+    NUM_MONSTERS = (30, 40)
+
+    NUM_ROOMS = (15, 30)
+    ROOM_MAX_SIZE = 10
+    ROOM_MIN_SIZE = 6
+    NUM_TREES = 1800
+    MAX_ROOM_MONSTERS = 3
+    MAX_ROOM_ITEMS = 2
+
+    def __init__(self, area_map):
+        self._rooms = []
+        self._area_map = area_map
+        self._generate_rooms()
+        self._generate_objects()
+
+    def _generate_rooms(self):
+        # TODO: dry this block with forest generator
+        for x in range(0, self._area_map.width):
+            for y in range(0, self._area_map.height):
+                self._convert_to_wall(self._area_map.tiles[x][y])
+
+        rooms_to_generate = random.randint(DungeonGenerator.NUM_ROOMS[0], DungeonGenerator.NUM_ROOMS[1])
+
+        # If you generate a room overlapping another room, that's a fail.
+        # After ten failures, we give up and return the dungeon as-is.
+        num_failures = 10
+
+        while rooms_to_generate and num_failures:
+            # random width and height
+            w = randint(DungeonGenerator.ROOM_MIN_SIZE, DungeonGenerator.ROOM_MAX_SIZE)
+            h = randint(DungeonGenerator.ROOM_MIN_SIZE, DungeonGenerator.ROOM_MAX_SIZE)
+            # random position without going out of the boundaries of the map
+            x = randint(0, self._area_map.width - w - 1)
+            y = randint(0, self._area_map.height - h - 1)
+
+            # "Rect" class makes rectangles easier to work with
+            new_room = Rect(x, y, w, h)
+
+            # run through the other rooms and see if they intersect with this one
+            failed = False
+            for other_room in self._rooms:
+                if new_room.intersect(other_room):
+                    failed = True
+                    num_failures -= 1
+                    break
+
+            if not failed:
```

Comment:
Probably irrelevant syntactic sugar, but you could probably get rid of the `failed` control variable with the `for else` construct.

Reply:
I haven't learned `for/else` yet, that's coming up in my course inshaAllah. Feel free to tweak it yourself since you're the Python expert here :)